KAFKA-12382: add a README for KIP-500 (#10227)
apache/kafka: KIP-500.md

KIP-500 Early Access Release
============================

# Introduction
It is now possible to run Apache Kafka without Apache ZooKeeper!  We call this mode [self-managed mode](https://cwiki.apache.org/confluence/display/KAFKA/KIP-500%3A+Replace+ZooKeeper+with+a+Self-Managed+Metadata+Quorum).  It is currently *EARLY ACCESS AND SHOULD NOT BE USED IN PRODUCTION*, but it is available for testing in the Kafka 2.8 release.

When the Kafka cluster is in self-managed mode, it does not store its metadata in ZooKeeper.  In fact, you do not have to run ZooKeeper at all, because it stores its metadata in a Raft quorum of controller nodes.

Self-managed mode has many benefits -- some obvious, and some not so obvious.  Clearly, it is nice to manage and configure one service rather than two services.  In addition, you can now run a single process Kafka cluster.  Most important of all, self-managed mode is more scalable.  We expect to be able to [support many more topics and partitions](https://www.confluent.io/kafka-summit-san-francisco-2019/kafka-needs-no-keeper/) in this mode.

# Quickstart

## Warning

> Self-managed mode in Kafka 2.8 is provided for testing only, *NOT* for production.  We do not yet support upgrading existing ZooKeeper-based Kafka clusters into this mode.  In fact, when Kafka 3.0 is released, it may not even be possible to upgrade your self-managed clusters from 2.8 to 3.0 without downtime.  There may be bugs, including serious ones.  You should *assume that your data could be lost at any time* if you try the early access release of KIP-500.

Reviewer: > it may not even be possible to upgrade your self-managed clusters from 2.8 to 3.0 without downtime.

I think this statement as currently worded implies that we are committing to supporting an upgrade (potentially with some downtime).  We should drop the qualifier "without downtime" at the end if we are not committing to that.

Author: ok, I removed the "without downtime" part

## Generate a cluster ID

> The first step is to generate an ID for your new cluster, using the kafka-storage tool:

Reviewer 2: It would be useful to add a couple sentences of explanation for the clusterId generation and the storage tool since these are new. To keep the instructions concise here, maybe we could refer to a separate section which has more background.

Author: Hmm, I'm not sure what else to add here.  Can you give an example?

Reviewer 2: Why do we have this tool? Users will certainly ask that since these steps were not needed before. There are good reasons for it, so why don't we give a couple sentences of explanation?

~~~~
$ ./bin/kafka-storage.sh random-uuid
xtzWWN4bTjitpL3kfd9s5g
~~~~

## Format Storage Directories

> The next step is to format your storage directories.  If you are running in single-node mode, you can do this with one command:

Reviewer: Maybe use the term "combined" since that is what the prop file is called?  For example:

`If you are running in combined mode where you run the broker and controller in a single process, you can do this with one command:`

Author: I wanted to avoid talking about combined mode before the section on Deployment. However, I will change this to "For a single-process test cluster..." to keep things simple.

~~~~

> $ ./bin/kafka-storage.sh format -t xtzWWN4bTjitpL3kfd9s5g -c ./config/raft-combined.properties

Reviewer 3: If this is "self-contained mode" it's a bit confusing for the unitiated that the example configs ares `raft-*`. Not everyone reading this README will know the role of Raft in this. Perhaps the file should be `self-contained-combined.properties`, though that's maybe too long.

Author: Perhaps it could be "nozk-combined.properties", etc. just to keep it simple? I think that should be a separate PR, though (do you want to file it)?

Reviewer: How about `sm-{combined,broker,controller}.properties` where `sm` stands for `self-managed`?  I like that better than `nozk` becauae in a few years we will still have "zk" in the name; I think a positive name -- it's a config file for a self-managed case -- is better than a negative name (Look Mom, no ZooKeeper!)

Reviewer 3: @rondagostino the problem I see with that is it only makes sense if someone's told you what sm means, `nozk` is more self-explanatory for anyone who knows the 40,000ft view that Kafka is moving away from ZooKeeper. Are we really going to support the Kafka 3.x series for _years_? In my ignorance, I would have expected quite strong pressure to move on to Kafka 4.0 (i.e. no ZK support) fairly quickly (maybe 12 months after 3.0?) simply to avoid having to maintain two implementations of so much code.

Do we even need a prefix at all though? We already have `server.properties`=ZK, adding new `broker.properties`, `controller.properties` and `combined.properties` (or `controller+broker.properties`) seems to me to be self-describing and both backwards and forwards compatible. The only slight issue would be a newbie accidentally running with `broker.properties` rather than `server.properties`, but that would fail safe without a quorum.

Reviewer: > Do we even need a prefix at all though? We already have server.properties=ZK, adding new broker.properties, controller.properties and combined.properties (or controller+broker.properties) seems to me to be self-describing and both backwards and forwards compatible

I tend to agree with this suggestion.  I think using `{broker,controller,combined}.properties` for names would be a good way to go here.

Author: I agree that in the long term `{broker,controller,combined}.properties` are nice.  In the short term, though this could create confusion for people who want to run with ZK enabled (which is going to be most people, until we get out of "preview")  So let's go with nozk-* for now....

Formatting /tmp/raft-combined-logs
~~~~

> If you are using multiple nodes, then you should run the format command on each node.  Be sure to use the same cluster ID for each one.

Reviewer: Maybe it would be good to give the example since we provide the broker and controller configs?  Maybe this?

Be sure to use the same cluster ID for each one (e.g. run the format command against both `./config/raft-broker.properties` and `./config/raft-controller.properties` if you wish to run separate processes instead of a single one).

Author: It might be better to save that discussion for the "deployment section" since we haven't explained the different deployment models yet... or I guess we could move the deployment section to be first... not sure.

## Start the Kafka Server

> Finally, you are ready to start the Kafka server on each node.

Reviewer: ```suggestion
Finally, you are ready to start the Kafka server(s).
```

Author: I'll just say "server" (singular) since it is singular in this example.

[2021-02-26 15:37:11,466] INFO [Log partition=@metadata-0, dir=/tmp/nozk-combined-logs] Loading producer state till offset 0 with message format version 2 (kafka.log.Log)
[2021-02-26 15:37:11,509] INFO [raft-expiration-reaper]: Starting (kafka.raft.TimingWheelExpirationService$ExpiredOperationReaper)
[2021-02-26 15:37:11,640] INFO [RaftManager nodeId=1] Completed transition to Unattached(epoch=0, voters=[1], electionTimeoutMs=9037) (org.apache.kafka.raft.QuorumState)
...
~~~~

Just like with a ZooKeeper based broker, you can connect to port 9092 (or whatever port you configured) to perform administrative operations or produce or consume data.

~~~~
$ ./bin/kafka-topics.sh --create --topic foo --partitions 1 --replication-factor 1 --bootstrap-server localhost:9092
Created topic foo.
~~~~

# Deployment
Unlike in ZooKeeper-based mode, where any server can become the controller, in self-managed mode, only a small group of specially selected servers can act as controllers.  The specially selected controller servers will participate in the metadata quorum.  Each KIP-500 controller server is either active, or a hot standby for the current active controller server.

> Typically you will select either 3 or 5 servers for this role, depending on the size of your cluster.  Just like with ZooKeeper, you must keep a majority of the controllers alive in order to maintain availability.  So if you have 3 controllers, you can tolerate 1 failure; with 5 controllers, you can tolerate 2 failures.

Reviewer 2: I wonder if it's worth saying something about the relation between `process.roles`, `controller.quorum.voters`, and `controller.listeners`.

Author: Hmm, I'm not sure what to add...

Reviewer 2: I think the main thing is emphasizing that they need to be set consistently. We don't say anything about `quorum.voters` in this README currently, but it's important to understand the basics of how to configure it to run a multi-node setup.

Author: I added a section about `quorum.voters`

Each Kafka server now has a new configuration key called `process.roles` which can have the following values:

* If `process.roles` is set to `broker`, the server acts as a self-managed broker.
* If `process.roles` is set to `controller`, the server acts as a self-managed controller.
* If `process.roles` is set to `broker,controller`, the server acts as both a self-managed broker and a self-managd controller.
* If `process.roles` is not set at all then we are assumed to be in ZooKeeper mode.  As mentioned earlier, you can't yet transition back and forth between ZK mode and self-managed mode without reformatting.

> 

Reviewer 3: Here might be a good place to mention that `raft-broker.properties` and `raft-controller.properties` relate to the different `process.roles`, if Ron's earlier suggestion isn't accepted.

> Nodes that act as both brokers and controllers are referred to as "combined" nodes.  The advantage of using combined nodes you will have uses fewer Java Virtual Machines (JVMs).  This will allow you to avoid some of the fixed memory overheads associated with JVMs.  The disdavantage is that the controller will be less isolated from the rest of the system.  For example, if activity on the broker causes an out of memory condition, the controller part of the server is not isolated from that OOM condition. 

Reviewer: `s/The advantage of using combined nodes you will have uses fewer/One advantage of using combined nodes is that it uses fewer/`

# Missing Features
We do not yet support generating or loading KIP-630 metadata snapshots.  This means that after a while, the time required to restart a broker will become very large.  This is a known issue and we are working on implementing snapshots for the next release.

We also don't support any kind of upgrade right now, either to or from self-managed mode.  This is another big gap that we are working on.

Finally, the following Kafka features have not yet been fully implemented:

> * Support for security (configuring an Authorizer, setting up SCRAM, delegation tokens, and so forth)

Reviewer: I technically think it is possible to configure an authorizer (e.g. one that goes against some external datastore like LDAP -- though we haven't tested such a configuration).  But the lack of a tested configuration is perhaps reason to keep your wording as-is.

Maybe append the comment that the security protocols SASL_SSL and SASL_PLAINTEXT are supported for the non-SCRAM mechanisms and the SSL security protocol is also supported, but it is not yet possible to create ACLs for the authenticated identities?

Author: We disabled the authorizer APIs though for creating and listing ACLs, etc. So I would call it unsupported.

I will add a note that we support SSL, just to be clear.

* Support for transactions and exactly-once semantics
* Support for adding partitions to existing topics
* Support for partition reassignment
* Support for some configurations, like enabling unclean leader election by default or dynamically changing broker endpoints
* Support for KIP-112 "JBOD" modes
* Support for KIP-631 controller metrics

> We've tried to make it clear when a feature is not supported in the early access release, but you may encounter some rough edges.

Reviewer 3: Maybe mention whether/how to report such issues?

Author: I'll link to CONTRIBUTING.md here

KIP-500 Early Access Release
============================

# Introduction
It is now possible to run Apache Kafka without Apache ZooKeeper!  We call this mode [self-managed mode](https://cwiki.apache.org/confluence/display/KAFKA/KIP-500%3A+Replace+ZooKeeper+with+a+Self-Managed+Metadata+Quorum).  It is currently *EARLY ACCESS AND SHOULD NOT BE USED IN PRODUCTION*, but it is available for testing in the Kafka 2.8 release.

When the Kafka cluster is in self-managed mode, it does not store its metadata in ZooKeeper.  In fact, you do not have to run ZooKeeper at all, because it stores its metadata in a Raft quorum of controller nodes.

> Self-managed mode has many benefits-- some obvious, and some not so obvious.  Clearly, it is nice to manage and configure one service rather than two services.  In addition, you can now run a single process Kafka cluster.  Most important of all, self-managed mode is more scalable.  We expect to be able to [support many more topics and partitions](https://www.confluent.io/kafka-summit-san-francisco-2019/kafka-needs-no-keeper/) in this mode.

Reviewer 3: ```suggestion
Self-managed mode has many benefits -- some obvious, and some not so obvious.  Clearly, it is nice to manage and configure one service rather than two services.  In addition, you can now run a single process Kafka cluster.  Most important of all, self-managed mode is more scalable.  We expect to be able to [support many more topics and partitions](https://www.confluent.io/kafka-summit-san-francisco-2019/kafka-needs-no-keeper/) in this mode.
```

## Process Roles
Each Kafka server now has a new configuration key called `process.roles` which can have the following values:

* If `process.roles` is set to `broker`, the server acts as a self-managed broker.
* If `process.roles` is set to `controller`, the server acts as a self-managed controller.
* If `process.roles` is set to `broker,controller`, the server acts as both a self-managed broker and a self-managd controller.
* If `process.roles` is not set at all then we are assumed to be in ZooKeeper mode.  As mentioned earlier, you can't yet transition back and forth between ZK mode and self-managed mode without reformatting.

Nodes that act as both brokers and controllers are referred to as "combined" nodes.  The advantage of using combined nodes you will have uses fewer Java Virtual Machines (JVMs).  This will allow you to avoid some of the fixed memory overheads associated with JVMs.  The disdavantage is that the controller will be less isolated from the rest of the system.  For example, if activity on the broker causes an out of memory condition, the controller part of the server is not isolated from that OOM condition.

## Quorum Voters

> All nodes in the system must set the `quorum.voters` configuration.  This identifies the quorum controller servers that should be used.  All the controllers must be enumerated.  This is similar to how, when using ZooKeeper, the `zookeeper.connect` configuration must contain all the ZooKeeper servers.  Unlike with the ZK config, however, `quorum.voters` also has IDs for each node.  The format is id1@host1:port1,id2@host2:port2, etc.

Reviewer 2: Should be `controller.quorum.voters`

Author: good catch

So if you have 10 brokers and 3 controllers named controller1, controller2, controller3, you might have the following `quorum.voters` configuration:
```
1@controller1.example.com:9093,2@controller2.example.com:9093,3@controller3.example.com:9093

> ```

Reviewer 2: Can you include an example of the listener configuration. Just something which emphasizes that port 9093 is defined as the controller listener. For example:
```
node.id=1
listeners=CONTROLLER://controller1.example.com:9093
controller.listener.name=CONTROLLER
```
